finesse/hardware/temperature/dp9800.py

```diff
@@ -0,0 +1,219 @@
+"""This module provides an interface to DP9800 temperature readers."""
+import logging
+from datetime import datetime
+from decimal import Decimal
+
+from pubsub import pub
+from serial import Serial, SerialException, SerialTimeoutException
+
+
+def check_data(data: bytes) -> None:
+    """Perform message integrity checks.
+
+    First check characters that we know should be constant.
+    Then calculate the BCC and compare it to that transmitted.
+
+    Args:
+        data: the message to check
+
+    Raises:
+        DP9800Error: Malformed message received from device
+    """
+    header_width = 2
+    value_width = 8
+    num_values = 9
+    if len(data) < (header_width + value_width * num_values):
+        raise DP9800Error("Insufficient data read")
+
+    if data[0] != 2:  # STX
+        raise DP9800Error("Start transmission character not detected")
+    if data.find(3) == -1:  # ETX
+        raise DP9800Error("End transmission character not detected")
+    if data[-1] != 0:  # NUL
+        raise DP9800Error("Null terminator not detected")
+
+    bcc = data[-2]
+    bcc_chars = data[1:-2]
+    byte_sum = 0
+    for byte in bcc_chars:
+        byte_sum ^= byte
+
+    if byte_sum != bcc:
+        raise DP9800Error("BCC check failed")
+
+
+def parse_data(data: bytes) -> tuple[list[Decimal], str]:
```

Comment:
I'd call  `check_data()` directly from this function. I don't think it makes sense to have checking the data be a separate step that the user has to remember to do.